### PR TITLE
Try to add pytorch

### LINKED ISSUE #312: Add pytorch to ML image (or create separate image)
For LEAP and M2LInES 2i2c Hubs, many researchers will want to use pytorch. Unfortunately, this image does not have pytorch.

When I try to add pytorch-gpu to `ml-notebook/environment.yml` and call

```
conda-lock lock --mamba -f environment.yml -f ../pangeo-notebook/environment.yml -p linux-64
```

I get the error

```
Encountered problems while solving.
Problem: nothing provides __glibc >=2.17,<3.0.a0 needed by pytorch-1.10.0-cuda102py37h689c94d_1
```

I found a similar problem in https://github.com/NVIDIA/MinkowskiEngine/issues/354. That issue implies we may need to mix conda channels to get this to work. The [workaround](https://github.com/NVIDIA/MinkowskiEngine/issues/354#issuecomment-880555195) they recommend is to install openblas from `anaconda` and then switch to conda-forge for pytorch.

Any thoughts on the best path here?

cc @yuvipanda  

Also related to #188

### PR DESCRIPTION
**Pull request by rabernat**
closes #312

### CODE REVIEW

**Comment by rabernat** [comment]
Whoa it actually worked!

File: ml-notebook/packages.txt [hunk 1]
```diff
@@ -383,8 +390,9 @@ netcdf-fortran==4.5.4
 netcdf4==1.5.8
 nettle==3.6
 networkx==2.8
+ninja==1.10.2
 noise==1.2.2
-nomkl==1.0
+nomkl==2.0
```
Comment (scottyhq) [inline ml-notebook/packages.txt:395]: odd to see both 'mkl' and 'nomkl'

File: pangeo-notebook/environment.yml [hunk 2]
```diff
@@ -3,6 +3,9 @@ channels:
  - conda-forge
 dependencies:
  - python=3.9*
+ # needed for pytorch to work
+ # https://github.com/NVIDIA/MinkowskiEngine/issues/354#issuecomment-880555195
+ - anaconda::libopenblas
```
[the remark this answers is not in the document]
Comment (scottyhq) [reply pangeo-notebook/environment.yml:8]: we used to do this for installing various nvidia libraries, i guess it's okay, although i notice our logs current don't seem to show which repository packages come from...

File: ml-notebook/packages.txt [hunk 3]
```diff
@@ -350,6 +356,7 @@ matplotlib-inline==0.1.3
 mercantile==1.2.1
 metpy==1.3.0
 mistune==0.8.4
+mkl==2022.0.1
```
Comment (scottyhq) [inline ml-notebook/packages.txt:359]: this appears in addition to 'nomkl' package

File: ml-notebook/packages.txt (hunk 1 above)
Comment (scottyhq) [reply ml-notebook/packages.txt:395]: nomkl is coming from https://github.com/pangeo-data/pangeo-docker-images/blob/b5c38d09ca44e66dfda529fbfeb15ba4789a3805/pangeo-notebook/environment.yml#L58, presumably mkl is coming with pytorch...

**Comment by scottyhq** [comment]
I'm surprised this worked. The main concern I have is added complexity and huge image size (it seems this takes us from 6.63GB ->  8.48GB). 

Does anyone actually use pytorch and tensorflow together? I've always seen one or the other. Planetary computer images break them into 2 (https://github.com/microsoft/planetary-computer-containers), should we do the same?

**Comment by rabernat** [comment]
No, I don't think anyone uses them together. It's an either / or situation. Would you prefer to separate them into distinct images?

**Comment by scottyhq** [comment]
> No, I don't think anyone uses them together. It's an either / or situation. Would you prefer to separate them into distinct images? 

Yes, I think that would be better

**Comment by rabernat** [comment]
So we would rename
- `pangeo-ml` -> `pangeo-tensorflow`
and create a new image
- `pangeo-torch`

The only downside of splitting is that it requires a bit of refactoring to our CI. This is not something I will have the bandwidth for in the near term. @yuvipanda - given that 2i2c is using these images heavily for three separate hubs, is there a chance that someone from your team could implement this change?

**Comment by weiji14** [comment]
Hi there, mind if I take a crack at creating a `pangeo-torch` image? I can start a separate PR for this. Is it just `torch` and `torchgeo` that is needed, or can I throw in a couple of other packages as well like `pytorch-lightning` (to mimic https://github.com/microsoft/planetary-computer-containers/blob/2022.04.11.1/gpu-pytorch/environment.yml)?

**Comment by rabernat** [comment]
Welcome @weiji14! We would be happy to have your help with this!

The main challenge for an outside contributor is to understand the CI we use to build these images. We try to document it, but I'm sure there will be some areas you find confusing. So please give it a try, but don't hesitate to ask for help!

**Comment by weiji14** [comment]
Cool, let me see if I can find my way around the GitHub Actions CI :laughing: By the way, is it ok to name the docker image `pytorch-notebook` instead of `torch-notebook`?

**Comment by rabernat** [comment]
Sure, `pytorch-notebook` and `tensorflow-notebook`.

**Comment by weiji14** [comment]
Ok, PR started at #315. Maybe do the `ml-notebook` -> `tensorflow-notebook` rename in a separate Pull Request to keep things clean.

**Comment by scottyhq** [comment]
closed in favor of #315

**Pull request closed without merging**